src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsStream.cs

```diff
@@ -0,0 +1,168 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Data.SqlClientX.IO
+{
+    /// <summary>
+    /// A Stream abstraction over the TDS protocol.
+    /// The stream can be used to read and write TDS messages on a 
+    /// SQL Server physical connection.
+    /// </summary>
+    internal class TdsStream : Stream
```

Comment:
It would make sense to encapsulate the additional APIs, which are not part of stream, in interfaces.

Reply:
This is where I was getting hung up. Whatever interface we'd create (let's just say `ITdsStream` for this example), should inherit from Stream, but Stream doesn't have an interface. If we want the consumer of this class to be able to substitute in other `Stream`s, then the consumer won't be able to rely on the additional methods (unless it constantly does a check for the interface before trying to call the additional method, then falling back to the other if it doesn't work, which is not ideal at all). If the consumer takes a dependency on `TdsStream`, then they can safely rely on the additional methods, but won't be able to use mocks (unless we make the additional methods virtual).

So, my suggestion here is:
* Make the additional methods virtual, so that they can be mocked by consumers
* *If* there will be more than once class providing these methods, we could add an interface for them, just to ensure that the classes implement them the same way.

Reply:
Thanks, this makes sense. 

I was going to talk to you about this, and you have given me the answer

Reply:
Another benefit of the second suggestion ... extension methods. Eg, we could offer the stream extensions that write numbers to the stream via multiple calls to WriteByteAsync, it could be an extension method of `ITdsStream`. That way we don't have to allocate buffers for 4 bytes of int, just to copy it into the stream. We could instead do 4 separate calls to write one byte at a time.